From 812c5e58bd743ea26d3dec8d668466535be8b661 Mon Sep 17 00:00:00 2001
From: Weili Qian <qianweili@huawei.com>
Date: Tue, 7 Feb 2023 09:42:34 +0800
Subject: [PATCH 890/905] [Backport] crypto: hisilicon/qm - add functions for
 releasing resources

mainline inclusion
from mainline-v6.0-rc1
commit bf081d6fa8e90aefe991e34a29eff7aa22deb3ff
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I6DRLU

Reference: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=bf081d6fa8e9

----------------------------------------------------------------------

The resources allocated by hisi_qm_memory_init() are released by
hisi_qm_uninit(). Add hisi_qm_memory_uninit() to release resources,
no functional change.

conficts:
	drivers/crypto/hisilicon/qm.c

Signed-off-by: Weili Qian <qianweili@huawei.com>
Signed-off-by: JiangShui Yang <yangjiangshui@h-partners.com>
Signed-off-by: Keyi Zhong <zhongkeyi1@huawei.com>
---
 drivers/crypto/hisilicon/qm.c | 30 +++++++++++++++++-------------
 1 file changed, 17 insertions(+), 13 deletions(-)

diff --git a/drivers/crypto/hisilicon/qm.c b/drivers/crypto/hisilicon/qm.c
index 57de083e8967..ef09e45bdc3e 100644
--- a/drivers/crypto/hisilicon/qm.c
+++ b/drivers/crypto/hisilicon/qm.c
@@ -3480,6 +3480,21 @@ static void hisi_qm_pci_uninit(struct hisi_qm *qm)
 	pci_disable_device(pdev);
 }
 
+static void hisi_qm_memory_uninit(struct hisi_qm *qm)
+{
+	struct device *dev = &qm->pdev->dev;
+
+	hisi_qp_memory_uninit(qm, qm->qp_num);
+	if (qm->qdma.va) {
+		hisi_qm_cache_wb(qm);
+		dma_free_coherent(dev, qm->qdma.size,
+				  qm->qdma.va, qm->qdma.dma);
+	}
+
+	idr_destroy(&qm->qp_idr);
+	kfree(qm->factor);
+}
+
 /**
  * hisi_qm_uninit() - Uninitialize qm.
  * @qm: The qm needed uninit.
@@ -3488,26 +3503,15 @@ static void hisi_qm_pci_uninit(struct hisi_qm *qm)
  */
 void hisi_qm_uninit(struct hisi_qm *qm)
 {
-	struct pci_dev *pdev = qm->pdev;
-	struct device *dev = &pdev->dev;
-
 	qm_cmd_uninit(qm);
-	kfree(qm->factor);
-	down_write(&qm->qps_lock);
 
+	down_write(&qm->qps_lock);
 	if (!qm_avail_state(qm, QM_CLOSE)) {
 		up_write(&qm->qps_lock);
 		return;
 	}
 
-	hisi_qp_memory_uninit(qm, qm->qp_num);
-	idr_destroy(&qm->qp_idr);
-
-	if (qm->qdma.va) {
-		hisi_qm_cache_wb(qm);
-		dma_free_coherent(dev, qm->qdma.size,
-				  qm->qdma.va, qm->qdma.dma);
-	}
+	hisi_qm_memory_uninit(qm);
 	up_write(&qm->qps_lock);
 
 	hisi_qm_set_state(qm, VF_NOT_READY);
-- 
2.17.1

